samza-test/src/test/java/org/apache/samza/test/processor/TestZkStreamProcessor.java

```diff
@@ -0,0 +1,558 @@
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.samza.config.ApplicationConfig;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.JobCoordinatorConfig;
+import org.apache.samza.config.MapConfig;
+import org.apache.samza.config.ZkConfig;
+import org.apache.samza.processor.StreamProcessor;
+import org.apache.samza.processor.StreamProcessorLifecycleListener;
+import org.apache.samza.system.IncomingMessageEnvelope;
+import org.apache.samza.system.OutgoingMessageEnvelope;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.task.InitableTask;
+import org.apache.samza.task.MessageCollector;
+import org.apache.samza.task.StreamTask;
+import org.apache.samza.task.TaskContext;
+import org.apache.samza.task.TaskCoordinator;
+import org.apache.samza.test.StandaloneIntegrationTestHarness;
+import org.apache.samza.test.StandaloneTestUtils;
+import org.apache.samza.zk.TestZkUtils;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestZkStreamProcessor extends StandaloneIntegrationTestHarness {
+  /**
+   * Testing a basic identity stream task - reads data from a topic and writes it to another topic
+   * (without any modifications)
+   *
+   * <p>
+   * The standalone version in this test uses KafkaSystemFactory and it uses a SingleContainerGrouperFactory. Hence,
+   * no matter how many tasks are present, it will always be run in a single processor instance. This simplifies testing
+   */
+
+  public static StreamProcessorLifecycleListener listener = new StreamProcessorLifecycleListener() {
+    @Override
+    public void onStart() {
+    }
+
+    @Override
+    public void onShutdown() {
+    }
+
+    @Override
+    public void onFailure(Throwable t) {
+    }
+  };
+
+  @Test
+  public void testSingleStreamProcessor() {
+    testStreamProcessor(new String[]{"1"});
+  }
+
+  @Test
+  public void testTwoStreamProcessors() {
+    testStreamProcessor(new String[]{"1", "2"});
+  }
+
+  @Test
+  public void testFiveStreamProcessors() {
+    testStreamProcessor(new String[]{"1", "2", "3", "4", "5"});
+  }
+
+  // main test method for happy path with fixed number of processors
+  private void testStreamProcessor(String[] processorIds) {
+    final String testSystem = "test-system";
+    final String inputTopic = "numbers";
+    final String outputTopic = "output";
+    final int messageCount = 40;
+
+    final Map<String, String> map = createConfigs(testSystem, inputTopic, outputTopic, messageCount);
+
+    // Note: createTopics needs to be called before creating a StreamProcessor. Otherwise it fails with a
+    // TopicExistsException since StreamProcessor auto-creates them.
+    createTopics(inputTopic, outputTopic);
+
+    // create a latch of the size == number of messages
+    TestStreamTask.endLatch = new CountDownLatch(messageCount);
+
+    StreamProcessor[] streamProcessors = new StreamProcessor[processorIds.length];
```

Comment:
well, the tests require some much more finer control, which is not generally available through the applicaionRunner.